datajunction-server/datajunction_server/sql/parsing/ast.py

```diff
@@ -2640,6 +2650,74 @@ async def compile(self, ctx: CompileContext):
         if self._is_compiled:
             return
 
+        def _compile(info: Tuple[Column, List[TableExpression]]):
+            """
+            Given a list of table sources, find a matching origin table for the column.
+            """
+            col, table_options = info
+            matching_origin_tables = 0
+            for option in table_options:
+                namespace = col.namespace[0].name if col.namespace else None
+                table_alias = option.alias.name if option.alias else None
+                if not namespace or (namespace and namespace == table_alias):
+                    result = option.add_reference_column(col)
+                    if result:
+                        matching_origin_tables += 1
+                        col._is_compiled = True
+            if matching_origin_tables > 1:
+                ctx.exception.errors.append(
+                    DJError(
+                        code=ErrorCode.INVALID_COLUMN,
+                        message=f"Column `{col.name.name}` found in multiple tables."
+                        " Consider using fully qualified name.",
+                    ),
+                )
+
+        # Work backwards from the table expressions on the query's SELECT clause
+        # and assign references between the columns and the tables
+        nearest_query = self.get_nearest_parent_of_type(Query)
+        cte_mapping = {
+            cte.alias_or_name.name: cte
+            for cte in (nearest_query.ctes if nearest_query else [])
+        }
+        table_options = (
+            list(self.select.from_.find_all(TableExpression))
+            if self.select.from_
+            else []
+        )
+        if table_options:
+            for idx, option in enumerate(table_options):
+                if isinstance(option, Table):
+                    if option.name.name in cte_mapping:
+                        table_options[idx] = cte_mapping[option.name.name]
+                await table_options[idx].compile(ctx)
+
+            expressions_to_compile = [
+                self.select.projection,
+                self.select.group_by,
+                self.select.having,
+                self.select.where,
+                self.select.lateral_views,
+                self.select.set_op,
+                self.select.organization,
+            ]
+            columns_to_compile = []
+            for expression in expressions_to_compile:
+                if expression and not isinstance(expression, list):
+                    columns_to_compile += list(expression.find_all(Column))
+                if isinstance(expression, list):
+                    columns_to_compile += [
+                        col for expr in expression for col in expr.find_all(Column)
+                    ]
+
+            with ThreadPoolExecutor() as executor:
```

Comment:
Oh interesting... this should def speed things up. But I wonder if these executions can still use some partial caching or is every one of them going to repeat the same work for columns originating from the same table. I guess it depend what optimizations are done in `_compile()`.

Reply:
The primary optimization from the changes to `_compile` is the fact that we now start with a parent that already exists, so we're not going to every child and then traversing up the tree to find the parents, which actually consumes quite a bit of time.